File: recipe_catalog.py


### 1. `recipe_catalog.py` (Python)

```python
# recipe_catalog.py — Python версия

import json
import os
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)
DATA_FILE = "recipes.json"

class Recipe:
    def __init__(self, id, name, category, ingredients, instructions, prep_time, difficulty, rating=0):
        self.id = id
        self.name = name
        self.category = category
        self.ingredients = ingredients
        self.instructions = instructions
        self.prep_time = prep_time
        self.difficulty = difficulty
        self.rating = rating
        self.ratings_count = 0

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "category": self.category,
            "ingredients": self.ingredients,
            "instructions": self.instructions,
            "prep_time": self.prep_time,
            "difficulty": self.difficulty,
            "rating": self.rating,
            "ratings_count": self.ratings_count
        }

    @classmethod
    def from_dict(cls, data):
        recipe = cls(data["id"], data["name"], data["category"], data["ingredients"],
                     data["instructions"], data["prep_time"], data["difficulty"], data.get("rating", 0))
        recipe.ratings_count = data.get("ratings_count", 0)
        return recipe

    def add_rating(self, rating):
        total = self.rating * self.ratings_count + rating
        self.ratings_count += 1
        self.rating = total / self.ratings_count

class RecipeCatalog:
    def __init__(self):
        self.recipes = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            try:
                with open(DATA_FILE, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    self.recipes = [Recipe.from_dict(r) for r in data]
            except:
                self.recipes = []

    def save(self):
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump([r.to_dict() for r in self.recipes], f, indent=2, ensure_ascii=False)

    def add(self, name, category, ingredients, instructions, prep_time, difficulty):
        id = len(self.recipes) + 1
        recipe = Recipe(id, name, category, ingredients, instructions, prep_time, difficulty)
        self.recipes.append(recipe)
        self.save()
        return id

    def list_all(self):
        if not self.recipes:
            print(Fore.YELLOW + "Каталог пуст.")
            return
        print(Fore.CYAN + f"{'ID':<4} {'Название':<25} {'Категория':<12} {'Время':<8} {'Сложность':<10} {'Рейтинг':<8}")
        print("-" * 75)
        for r in self.recipes:
            rating_str = f"{r.rating:.1f}★" if r.ratings_count > 0 else "—"
            diff_color = Fore.GREEN if r.difficulty == "лёгкая" else Fore.YELLOW if r.difficulty == "средняя" else Fore.RED
            print(f"{r.id:<4} {r.name[:25]:<25} {r.category[:12]:<12} {r.prep_time:<8} {diff_color}{r.difficulty[:10]:<10}{Style.RESET_ALL} {rating_str:<8}")

    def search(self, query):
        query = query.lower()
        results = [r for r in self.recipes if query in r.name.lower() or
                   any(query in ing.lower() for ing in r.ingredients)]
        if not results:
            print(Fore.YELLOW + "Ничего не найдено.")
            return
        for r in results:
            print(f"{r.id}: {r.name} | {r.category} | {r.prep_time} мин | {r.difficulty} | {r.rating:.1f}★")

    def filter_by(self, field, value):
        results = []
        for r in self.recipes:
            if field == "category" and r.category.lower() == value.lower():
                results.append(r)
            elif field == "difficulty" and r.difficulty.lower() == value.lower():
                results.append(r)
            elif field == "prep_time" and r.prep_time <= int(value):
                results.append(r)
        if not results:
            print(Fore.YELLOW + "Нет рецептов, соответствующих фильтру.")
            return
        for r in results:
            print(f"{r.id}: {r.name} | {r.category} | {r.prep_time} мин | {r.difficulty}")

    def sort_by(self, field, reverse=False):
        if field == "name":
            self.recipes.sort(key=lambda r: r.name, reverse=reverse)
        elif field == "prep_time":
            self.recipes.sort(key=lambda r: r.prep_time, reverse=reverse)
        elif field == "rating":
            self.recipes.sort(key=lambda r: r.rating, reverse=reverse)
        else:
            print(Fore.RED + "Неверное поле для сортировки.")
            return
        self.list_all()

    def delete(self, id):
        for i, r in enumerate(self.recipes):
            if r.id == id:
                del self.recipes[i]
                self.save()
                return True
        return False

    def edit(self, id, field, value):
        for r in self.recipes:
            if r.id == id:
                if field == "name":
                    r.name = value
                elif field == "category":
                    r.category = value
                elif field == "ingredients":
                    r.ingredients = [i.strip() for i in value.split(',')]
                elif field == "instructions":
                    r.instructions = value
                elif field == "prep_time":
                    r.prep_time = int(value)
                elif field == "difficulty":
                    r.difficulty = value
                else:
                    return False
                self.save()
                return True
        return False

    def rate(self, id, rating):
        for r in self.recipes:
            if r.id == id:
                r.add_rating(rating)
                self.save()
                return True
        return False

    def stats(self):
        if not self.recipes:
            print("Нет данных.")
            return
        total = len(self.recipes)
        categories = {}
        difficulties = {}
        total_rating = 0
        rated_count = 0
        for r in self.recipes:
            categories[r.category] = categories.get(r.category, 0) + 1
            difficulties[r.difficulty] = difficulties.get(r.difficulty, 0) + 1
            if r.ratings_count > 0:
                total_rating += r.rating
                rated_count += 1
        avg_rating = total_rating / rated_count if rated_count > 0 else 0
        print(Fore.CYAN + "📊 Статистика:")
        print(f"  Всего рецептов: {total}")
        print(f"  Средний рейтинг: {avg_rating:.1f}★ (из {rated_count} оценённых)")
        print("  По категориям:")
        for c, count in sorted(categories.items(), key=lambda x: -x[1]):
            print(f"    {c}: {count}")
        print("  По сложности:")
        for d, count in difficulties.items():
            print(f"    {d}: {count}")

    def export_json(self, filename="recipes_export.json"):
        data = [r.to_dict() for r in self.recipes]
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(Fore.GREEN + f"💾 Экспорт JSON: {filename}")

    def export_csv(self, filename="recipes_export.csv"):
        if not self.recipes:
            print(Fore.YELLOW + "Нет данных для экспорта.")
            return
        import csv
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(["ID", "Название", "Категория", "Ингредиенты", "Инструкция", "Время", "Сложность", "Рейтинг"])
            for r in self.recipes:
                writer.writerow([r.id, r.name, r.category, ", ".join(r.ingredients), r.instructions, r.prep_time, r.difficulty, f"{r.rating:.1f}"])
        print(Fore.GREEN + f"💾 Экспорт CSV: {filename}")

def main():
    catalog = RecipeCatalog()
    while True:
        print(Fore.CYAN + "\n🍰 Dessert Recipe Catalog (Python)")
        print("1. Добавить рецепт")
        print("2. Показать все рецепты")
        print("3. Поиск рецептов")
        print("4. Фильтрация")
        print("5. Сортировка")
        print("6. Удалить рецепт")
        print("7. Редактировать рецепт")
        print("8. Оценить рецепт")
        print("9. Статистика")
        print("10. Экспорт")
        print("11. Выход")
        choice = input("Выберите действие: ").strip()
        if choice == "1":
            name = input("Название: ")
            category = input("Категория (торт, пирожное, печенье, мороженое, другое): ").strip()
            ingredients = [i.strip() for i in input("Ингредиенты (через запятую): ").split(',')]
            instructions = input("Инструкция: ")
            prep_time = int(input("Время приготовления (мин): "))
            difficulty = input("Сложность (лёгкая/средняя/сложная): ").strip().lower()
            if difficulty not in ["лёгкая", "средняя", "сложная"]:
                difficulty = "средняя"
            id = catalog.add(name, category, ingredients, instructions, prep_time, difficulty)
            print(Fore.GREEN + f"✅ Рецепт добавлен (ID: {id})")
        elif choice == "2":
            catalog.list_all()
        elif choice == "3":
            query = input("Введите запрос (название или ингредиент): ")
            catalog.search(query)
        elif choice == "4":
            print("Фильтровать по: category, difficulty, prep_time")
            field = input("Поле: ").lower()
            value = input("Значение: ")
            if field == "prep_time":
                value = int(value)
            catalog.filter_by(field, value)
        elif choice == "5":
            print("Сортировать по: name, prep_time, rating")
            field = input("Поле: ").lower()
            rev = input("По убыванию? (y/n): ").lower() == 'y'
            catalog.sort_by(field, rev)
        elif choice == "6":
            catalog.list_all()
            id = int(input("Введите ID для удаления: "))
            if catalog.delete(id):
                print(Fore.GREEN + "✅ Рецепт удалён.")
            else:
                print(Fore.RED + "❌ Рецепт не найден.")
        elif choice == "7":
            catalog.list_all()
            id = int(input("Введите ID для редактирования: "))
            field = input("Какое поле редактировать (name, category, ingredients, instructions, prep_time, difficulty): ").lower()
            value = input("Новое значение: ")
            if catalog.edit(id, field, value):
                print(Fore.GREEN + "✅ Рецепт обновлён.")
            else:
                print(Fore.RED + "❌ Не удалось обновить.")
        elif choice == "8":
            catalog.list_all()
            id = int(input("Введите ID для оценки: "))
            rating = int(input("Оценка (1-5): "))
            if 1 <= rating <= 5:
                if catalog.rate(id, rating):
                    print(Fore.GREEN + "✅ Оценка добавлена.")
                else:
                    print(Fore.RED + "❌ Рецепт не найден.")
            else:
                print(Fore.RED + "❌ Оценка должна быть от 1 до 5.")
        elif choice == "9":
            catalog.stats()
        elif choice == "10":
            print("1. Экспорт в JSON")
            print("2. Экспорт в CSV")
            sub = input("Выберите формат: ")
            if sub == "1":
                catalog.export_json()
            elif sub == "2":
                catalog.export_csv()
            else:
                print(Fore.RED + "Неверный выбор.")
        elif choice == "11":
            print("До свидания!")
            break
        else:
            print(Fore.RED + "Неверный выбор.")

if __name__ == "__main__":
    main()

File: recipe_catalog.rb
# recipe_catalog.rb — Ruby версия

require 'json'
require 'csv'

DATA_FILE = 'recipes.json'

class Recipe
  attr_accessor :id, :name, :category, :ingredients, :instructions, :prep_time, :difficulty, :rating, :ratings_count

  def initialize(id, name, category, ingredients, instructions, prep_time, difficulty, rating = 0, ratings_count = 0)
    @id = id
    @name = name
    @category = category
    @ingredients = ingredients
    @instructions = instructions
    @prep_time = prep_time
    @difficulty = difficulty
    @rating = rating
    @ratings_count = ratings_count
  end

  def to_h
    { id: @id, name: @name, category: @category, ingredients: @ingredients,
      instructions: @instructions, prep_time: @prep_time, difficulty: @difficulty,
      rating: @rating, ratings_count: @ratings_count }
  end

  def self.from_h(h)
    new(h[:id], h[:name], h[:category], h[:ingredients], h[:instructions],
        h[:prep_time], h[:difficulty], h[:rating] || 0, h[:ratings_count] || 0)
  end

  def add_rating(rating)
    total = @rating * @ratings_count + rating
    @ratings_count += 1
    @rating = total / @ratings_count.to_f
  end
end

class RecipeCatalog
  attr_reader :recipes

  def initialize
    @recipes = []
    load
  end

  def load
    if File.exist?(DATA_FILE)
      begin
        data = JSON.parse(File.read(DATA_FILE), symbolize_names: true)
        @recipes = data.map { |r| Recipe.from_h(r) }
      rescue
        @recipes = []
      end
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@recipes.map(&:to_h)))
  end

  def add(name, category, ingredients, instructions, prep_time, difficulty)
    id = @recipes.size + 1
    @recipes << Recipe.new(id, name, category, ingredients, instructions, prep_time, difficulty)
    save
    id
  end

  def list_all
    if @recipes.empty?
      puts "\e[33mКаталог пуст.\e[0m"
      return
    end
    printf "\e[36m%-4s %-25s %-12s %-8s %-10s %-8s\e[0m\n", "ID", "Название", "Категория", "Время", "Сложность", "Рейтинг"
    puts "-" * 75
    @recipes.each do |r|
      rating_str = r.ratings_count > 0 ? "#{r.rating.round(1)}★" : "—"
      diff_color = r.difficulty == "лёгкая" ? "\e[32m" : r.difficulty == "средняя" ? "\e[33m" : "\e[31m"
      printf "%-4d %-25s %-12s %-8d %s%-10s\e[0m %-8s\n", r.id, r.name, r.category, r.prep_time, diff_color, r.difficulty, rating_str
    end
  end

  def search(query)
    query = query.downcase
    results = @recipes.select do |r|
      r.name.downcase.include?(query) || r.ingredients.any? { |i| i.downcase.include?(query) }
    end
    if results.empty?
      puts "\e[33mНичего не найдено.\e[0m"
      return
    end
    results.each { |r| puts "#{r.id}: #{r.name} | #{r.category} | #{r.prep_time} мин | #{r.difficulty} | #{r.rating.round(1)}★" }
  end

  def filter_by(field, value)
    results = case field
    when 'category'
      @recipes.select { |r| r.category.casecmp(value).zero? }
    when 'difficulty'
      @recipes.select { |r| r.difficulty.casecmp(value).zero? }
    when 'prep_time'
      @recipes.select { |r| r.prep_time <= value.to_i }
    else
      []
    end
    if results.empty?
      puts "\e[33mНет рецептов, соответствующих фильтру.\e[0m"
      return
    end
    results.each { |r| puts "#{r.id}: #{r.name} | #{r.category} | #{r.prep_time} мин | #{r.difficulty}" }
  end

  def sort_by(field, reverse)
    case field
    when 'name'
      @recipes.sort_by! { |r| r.name }
    when 'prep_time'
      @recipes.sort_by! { |r| r.prep_time }
    when 'rating'
      @recipes.sort_by! { |r| r.rating }
    else
      puts "\e[31mНеверное поле для сортировки.\e[0m"
      return
    end
    @recipes.reverse! if reverse
    list_all
  end

  def delete(id)
    found = @recipes.find { |r| r.id == id }
    if found
      @recipes.delete(found)
      save
      true
    else
      false
    end
  end

  def edit(id, field, value)
    recipe = @recipes.find { |r| r.id == id }
    return false unless recipe
    case field
    when 'name' then recipe.name = value
    when 'category' then recipe.category = value
    when 'ingredients' then recipe.ingredients = value.split(',').map(&:strip)
    when 'instructions' then recipe.instructions = value
    when 'prep_time' then recipe.prep_time = value.to_i
    when 'difficulty' then recipe.difficulty = value
    else return false
    end
    save
    true
  end

  def rate(id, rating)
    recipe = @recipes.find { |r| r.id == id }
    return false unless recipe
    recipe.add_rating(rating)
    save
    true
  end

  def stats
    if @recipes.empty?
      puts "Нет данных."
      return
    end
    categories = Hash.new(0)
    difficulties = Hash.new(0)
    total_rating = 0.0
    rated_count = 0
    @recipes.each do |r|
      categories[r.category] += 1
      difficulties[r.difficulty] += 1
      if r.ratings_count > 0
        total_rating += r.rating
        rated_count += 1
      end
    end
    avg_rating = rated_count > 0 ? total_rating / rated_count : 0
    puts "\e[36m📊 Статистика:\e[0m"
    puts "  Всего рецептов: #{@recipes.size}"
    puts "  Средний рейтинг: #{avg_rating.round(1)}★ (из #{rated_count} оценённых)"
    puts "  По категориям:"
    categories.sort_by { |_, c| -c }.each { |c, count| puts "    #{c}: #{count}" }
    puts "  По сложности:"
    difficulties.each { |d, count| puts "    #{d}: #{count}" }
  end

  def export_json(filename = 'recipes_export.json')
    File.write(filename, JSON.pretty_generate(@recipes.map(&:to_h)))
    puts "\e[32m💾 Экспорт JSON: #{filename}\e[0m"
  end

  def export_csv(filename = 'recipes_export.csv')
    if @recipes.empty?
      puts "\e[33mНет данных для экспорта.\e[0m"
      return
    end
    CSV.open(filename, 'w') do |csv|
      csv << ["ID", "Название", "Категория", "Ингредиенты", "Инструкция", "Время", "Сложность", "Рейтинг"]
      @recipes.each do |r|
        csv << [r.id, r.name, r.category, r.ingredients.join(', '), r.instructions, r.prep_time, r.difficulty, r.rating.round(1)]
      end
    end
    puts "\e[32m💾 Экспорт CSV: #{filename}\e[0m"
  end
end

def main
  catalog = RecipeCatalog.new
  loop do
    puts "\n\e[36m🍰 Dessert Recipe Catalog (Ruby)\e[0m"
    puts "1. Добавить рецепт"
    puts "2. Показать все рецепты"
    puts "3. Поиск рецептов"
    puts "4. Фильтрация"
    puts "5. Сортировка"
    puts "6. Удалить рецепт"
    puts "7. Редактировать рецепт"
    puts "8. Оценить рецепт"
    puts "9. Статистика"
    puts "10. Экспорт"
    puts "11. Выход"
    print "Выберите действие: "
    choice = gets.chomp
    case choice
    when "1"
      print "Название: "
      name = gets.chomp
      print "Категория (торт, пирожное, печенье, мороженое, другое): "
      category = gets.chomp
      print "Ингредиенты (через запятую): "
      ingredients = gets.chomp.split(',').map(&:strip)
      print "Инструкция: "
      instructions = gets.chomp
      print "Время приготовления (мин): "
      prep_time = gets.chomp.to_i
      print "Сложность (лёгкая/средняя/сложная): "
      difficulty = gets.chomp.downcase
      difficulty = "средняя" unless ["лёгкая", "средняя", "сложная"].include?(difficulty)
      id = catalog.add(name, category, ingredients, instructions, prep_time, difficulty)
      puts "\e[32m✅ Рецепт добавлен (ID: #{id})\e[0m"
    when "2"
      catalog.list_all
    when "3"
      print "Введите запрос (название или ингредиент): "
      query = gets.chomp
      catalog.search(query)
    when "4"
      puts "Фильтровать по: category, difficulty, prep_time"
      print "Поле: "
      field = gets.chomp.downcase
      print "Значение: "
      value = gets.chomp
      catalog.filter_by(field, value)
    when "5"
      puts "Сортировать по: name, prep_time, rating"
      print "Поле: "
      field = gets.chomp.downcase
      print "По убыванию? (y/n): "
      reverse = gets.chomp.downcase == 'y'
      catalog.sort_by(field, reverse)
    when "6"
      catalog.list_all
      print "Введите ID для удаления: "
      id = gets.chomp.to_i
      if catalog.delete(id)
        puts "\e[32m✅ Рецепт удалён.\e[0m"
      else
        puts "\e[31m❌ Рецепт не найден.\e[0m"
      end
    when "7"
      catalog.list_all
      print "Введите ID для редактирования: "
      id = gets.chomp.to_i
      print "Какое поле редактировать (name, category, ingredients, instructions, prep_time, difficulty): "
      field = gets.chomp.downcase
      print "Новое значение: "
      value = gets.chomp
      if catalog.edit(id, field, value)
        puts "\e[32m✅ Рецепт обновлён.\e[0m"
      else
        puts "\e[31m❌ Не удалось обновить.\e[0m"
      end
    when "8"
      catalog.list_all
      print "Введите ID для оценки: "
      id = gets.chomp.to_i
      print "Оценка (1-5): "
      rating = gets.chomp.to_i
      if (1..5).include?(rating)
        if catalog.rate(id, rating)
          puts "\e[32m✅ Оценка добавлена.\e[0m"
        else
          puts "\e[31m❌ Рецепт не найден.\e[0m"
        end
      else
        puts "\e[31m❌ Оценка должна быть от 1 до 5.\e[0m"
      end
    when "9"
      catalog.stats
    when "10"
      puts "1. Экспорт в JSON"
      puts "2. Экспорт в CSV"
      print "Выберите формат: "
      sub = gets.chomp
      if sub == "1"
        catalog.export_json
      elsif sub == "2"
        catalog.export_csv
      else
        puts "\e[31mНеверный выбор.\e[0m"
      end
    when "11"
      puts "До свидания!"
      break
    else
      puts "\e[31mНеверный выбор.\e[0m"
    end
  end
end

main if __FILE__ == $0
